cranelift/docs/isle-integration.md

```diff
@@ -0,0 +1,188 @@
+ISLE-generated code is glued into the rest of the system is with these trait
+implementations.
+
+There may also be a `lower` function defined in `isle.rs` that encapsulates
+creating the ISLE `Context` and calling into the generated code.
+
+## Lowering rules are always pure, use SSA
+
+The lowering rules themselves, defined in
+`cranelift/codegen/src/isa/<arch>/lower.isle`, must always be a pure mapping
+from a CLIF instruction to the target ISA's `MachInst`.
+
+Examples of things that the lowering rules themselves they shouldn't deal with
+or talk about:
+
+* Registers that are modified (both read and written to, violating SSA)
+* Implicit uses of registers
+* Maintaining use counts for each CLIF value or virtual register
+
+Instead, these things should be handled by some combination of
+`cranelift/codegen/src/isa/<arch>/inst.isle` and general Rust code (either in
+`cranelift/codegen/src/isa/<arch>/lower/isle.rs` or elsewhere).
+
+When an instruction modifies a register, both reading from it and writing to it,
+we should build an SSA view of that instruction that gets legalized via "move
+mitosis" by splitting a move out from the register.
+
+For example, on x86 the `add` instruction reads and writes its first operand:
+
+    add a, b    ==    a = a + b
+
+So we present an SSA facade where `add` operates on three registers, instead of
+two, and defines one of them, while reading the other two and leaving them
+unmodified:
+
+    add a, b, c    ==    a = b + c
+
+Then, as an implementation detail of the facade, we emit moves as necessary:
+
+    add a, b, c    ==>    mov a, b; add b, c
+
+We call the process of emitting these moves "move mitosis". For ISAs with
+ubiquitous use of modified registers and instructions in two-operand form, like
+x86, we implement move mitosis with methods on the ISA's `MachInst`. For other
+ISAs that are RISCier and where modified registers are pretty rare, such as
+aarch64, we implement the handful of move mitosis special cases at the
+`inst.isle` layer. Either way, the important thing is that the lowering rules
+remain pure.
+
+Fnally, note that these moves are generally cleaned up by the register
+allocator's move coalescing, and move mitosis will eventually go away completely
+once we switch over to `regalloc2`, which takes instructions in SSA form
+directly as input.
+
+Instructions that implicitly operate on specific registers, or which require
+that certain operands be in certain registers, are handled similarly: the
+lowering rules use a pure paradigm that ignores these constraints and has
+instructions that explicitly take implicit operands, and we ensure the
+constraints are fulfilled a layer below the lowering rules (in `inst.isle` or in
+Rust glue code).
+
+## When are lowering rules allowed to have side effects?
+
+Extractors (the matchers that appear on the left-hand sides of `rule`s) should
+**never** have side effects. When evaluating a rule's extractors, we haven't yet
```

Comment:
Is this/can this be enforced by the ISLE compiler? Or is it the responsibility of the rule-writer?

Reply:
The latter; the external extractor is a call into Rust, which could in theory do anything.

I had suggested at one point (in a conversation with @fitzgen I think) that we could split the `Context` trait into `InputContext` and `OutputContext`, and hold only an immutable borrow to `InputContext`, which would give us some type-level enforcement of this, at least...

Reply:
I thought some more about this and determined that if we actually did this, it would make rewriting our peephole optimizations in ISLE impossible because both the input and output contexts would need to borrow the DFG (shared/immutable borrow for input, exclusive/mutable for output) which would be a borrowck error. Or at least it would force us to use `RefCell` pervasively, which could be a perf hit and certainly would be a pain to work with.

Reply:
(Even the existing lowering doesn't have an input and output `LowerCtx` so we would have the same issues with instruction selection, even though we "shouldn't" if we were building this all from scratch)

Reply:
@fitzgen I think that the peephole (CLIF-to-CLIF) case is a separate design question with some interesting possibilities. The most natural approach (given our constraints) might actually be to send emitted instructions to a separate buffer (either a patch buffer on top of the IR with a later batch-update, or just emit new IR in a straight-through pass), in which case the Input and Output contexts are truly independent. That has the advantage of making the lowerings fine-grained-parallelizable as well. And it would allow us to move to a more VCode-like representation of "compact array of insts, no redirections or next/prev links" which is much more efficient to stream through. The downside is of course that it might take multiple passes to converge, but that's an effort-vs-quality tradeoff knob we can turn.

Anyway, that's a separate discussion (and one I'd love to dive into more!) but the relevance here is just to say "there's another way" :-)

Reply:
That would make it possible, yes. But it's also a lot further off than "porting simple preopt to ISLE" is, at least in my mind.

Reply:
Hmm, interesting -- we should definitely talk design more here, I think. I'm somewhat uncomfortable with the idea of a pass that does in-place edits, or at least it's not what I had in mind when designing the DSL or when thinking about the pipeline-of-streaming-passes architecture that the new backends are built around. My main concern is that it would paint us into a corner (in terms of rules that we write) where we depend on the immediate availability of the result of an edit, and can't get out of this corner later when we want to do any sort of batch-update optimization (separate streaming passes, or parallelization, or anything like that).

I think that it actually may not be so bad to buffer new instructions at insertion points and "replace value X with value Y" actions. In my estimation probably not any more work than what it would take to build in-place editing glue; both seem like ~a few hundred lines of code, in my estimation.

Anyway, we've branched off the thread topic here, and this is the sort of architectural thing that probably merits at least an issue thread of discussion, so I'm happy to discuss more elsewhere when it comes time for this!